packages/@pollyjs/adapter-http/src/utils/transport-wrapper.js

```diff
@@ -0,0 +1,198 @@
+
+const LISTENERS = Symbol();
+
+export default class TransportWrapper {
+  constructor(transport, { adapter }) {
+    this.adapter = adapter;
+    this.transport = transport;
+  }
+
+  patch() {
+    if (!this.transport[NATIVE_REQUEST]) {
+      this.transport[NATIVE_REQUEST] = this.transport.request;
+
+      this.transport.request = this.createRequestWrapper();
+    }
+  }
+
+  restore() {
+    if (this.transport[NATIVE_REQUEST]) {
+      this.transport.request = this.transport[NATIVE_REQUEST];
+
+      delete this.transport[NATIVE_REQUEST];
+    }
+  }
+
+  getHttpResponseData(res) {
+    return new Promise(resolve => {
+      let stream;
+
+      switch (res.headers['content-encoding']) {
+        case 'gzip':
+        case 'compress':
+        case 'deflate':
+          stream = res.pipe(zlib.createUnzip());
+          delete res.headers['content-encoding'];
+          break;
+        default:
+          stream = res;
+      }
+
+      const resBuffer = [];
+
+      stream.on('data', chunk => {
+        resBuffer.push(chunk);
+      });
+
+      stream.on('end', () => {
+        const data = Buffer.concat(resBuffer).toString('utf8');
+
+        resolve(data);
+      });
+    });
+  }
+
+  async passthrough(req) {
+    const res = await new Promise(resolve => {
+      req.once('response', response => resolve(response));
+      req[NATIVE_END].call(req);
+    });
+
+    const data = await this.getHttpResponseData(res);
+
+    const { headers, statusCode } = res;
+
+    return {
+      headers,
+      statusCode,
+      data
+    };
+  }
+
+  async respond(req, response) {
+    Object.keys(req[LISTENERS]).forEach(eventName => {
+      const listeners = req[LISTENERS][eventName];
+
+      listeners.forEach(listener => req.on(eventName, listener));
+    });
+
+    const fakeSocket = { readable: false };
+    const msg = new http.IncomingMessage(fakeSocket);
+
+    msg.statusCode = response.statusCode;
+    req.emit('response', msg);
+
+    for (const h in response.headers) {
+      msg.headers[h] = response.headers[h];
+    }
+
+    let body = response.body;
+
+    if (typeof body === 'undefined') {
+      body = [];
+    }
+
+    msg.emit('data', Buffer.from(body));
+    msg.emit('end');
+
+    req.emit('prefinish');
+    req.emit('finish');
+  }
+
+  createRequestWrapper() {
+    const wrapper = this;
+    const { adapter, transport } = wrapper;
+
+    return (...args) => {
+      /**
+       * args could be (url, options, callback) or (options, callback) depending on nodejs version
+       * We don't need to wrap callback we can just emit 'response' event on ClientRequest
+       * and it's more reliable to collect request information from ClientRequest instance
+       * We don't need to store args
+       */
+
+      const req = transport[NATIVE_REQUEST].call(transport, ...args);
+
+      const chunks = [];
+
+      const originalWrite = req.write;
+
+      req.write = (chunk, encoding, callback) => {
+        // TODO : handle encoding
```

Comment:
Is this still a todo?

Reply:
I'm still not sure how to handle binary content, i left it there to not forget it :)

Reply:
We're already discussing this in other discussion, i can remove this TODO comment now if you want ?